compiler/rustc_middle/src/ty/print/pretty.rs

```diff
@@ -641,104 +641,88 @@ pub trait PrettyPrinter<'tcx>:
+                    } else {
+                        p!(write("@{}", self.tcx().def_path_str(did)));
                     }
                 } else {
-                    p!(write("@{}", self.tcx().def_path_str(did)));
-
-                    if substs.as_generator().is_valid() {
-                        let upvar_tys = substs.as_generator().upvar_tys();
-                        let mut sep = " ";
-                        for (index, upvar_ty) in upvar_tys.enumerate() {
-                            p!(write("{}{}:", sep, index), print(upvar_ty));
-                            sep = ", ";
-                        }
+                    p!(print_def_path(did, substs));
+
+                    p!(write(" upvars_ty=("));
+                    let mut uninferred_ty = substs
+                        .as_closure()
+                        .upvar_tys()
+                        .filter(|ty| matches!(ty.kind(), ty::Infer(_)));
+                    if matches!(uninferred_ty.next(), Some(_)) {
+                        self = self.comma_sep(substs.as_generator().upvar_tys())?;
+                    } else {
+                        p!(write("unavailable"));
                     }
+                    p!(write(")"));
                 }
 
                 if substs.as_generator().is_valid() {
                     p!(write(" "), print(substs.as_generator().witness()));
                 }
 
-                p!(write("]"))
+                p!(write("]"));
             }
             ty::GeneratorWitness(types) => {
                 p!(in_binder(&types));
             }
             ty::Closure(did, substs) => {
-                p!(write("[closure"));
-
-                // FIXME(eddyb) should use `def_span`.
-                if let Some(did) = did.as_local() {
-                    let hir_id = self.tcx().hir().local_def_id_to_hir_id(did);
-                    if self.tcx().sess.opts.debugging_opts.span_free_formats {
-                        p!(write("@"), print_def_path(did.to_def_id(), substs));
-                    } else {
-                        let span = self.tcx().hir().span(hir_id);
-                        p!(write("@{}", self.tcx().sess.source_map().span_to_string(span)));
-                    }
-
-                    if substs.as_closure().is_valid() {
-                        let upvar_tys = substs.as_closure().upvar_tys();
-                        let mut sep = " ";
-                        for (&var_id, upvar_ty) in self
-                            .tcx()
-                            .upvars_mentioned(did)
-                            .as_ref()
-                            .iter()
-                            .flat_map(|v| v.keys())
-                            .zip(upvar_tys)
-                        {
-                            p!(write("{}{}:", sep, self.tcx().hir().name(var_id)), print(upvar_ty));
-                            sep = ", ";
+                p!(write("["));
+                if !self.tcx().sess.verbose() {
```

Comment:
I'd still like to see the parent substitutions in verbose mode - it would be very helpful for working on polymorphization. Perhaps a `-Z` flag could be added for this since most people probably don't need to see those?

Reply:
This is the non verbose case.

`p!(print_def_path(did, substs));` this would print parent substitution. 
Check https://github.com/rust-lang/rust/blob/343ba3879eca13aac9ed503baae6ba3a08214913/src/test/ui/closures/print/closure-print-generic-verbose-1.stderr

@nikomatsakis do we want a separate flag instead of -Zverbose?

Reply:
I'd prefer to have `-Zverbose` just print all the things, and not have to remember more flags.